### PR TITLE
Context sensitive dictation

### PR DESCRIPTION
**Pull request by rntz**
This implements context sensitive dictation, which determines how to space and capitalize dictated text by examining the text around the cursor before inserting.

I suggest reading this diff in the following order:

1. `modes/dictation_mode.talon`
2. `code/vocabulary.py`
3. `code/dictation.py`
4. `settings.talon`

The `word`, `text`, and `prose` captures got moved from vocabulary.py to dictation.py. This is because `text` & `prose` both need `format_text` and `prose` needs `auto_capitalize` and dictation needs `auto_capitalize`. We could split this by making something (eg. `auto_capitalize`) a talon action, but it seems like all this text-formatting code wants to be in the same place anyways. Open to other ways to organize this.

The capitalization code all became character based instead of word based, to avoid mangling whitespace. This makes it slightly more complex.

The PR leaves context sensitive dictation off by default, since, though useful, it is flaky and invasive; in some apps the copy-to-clipboard dance is unreliable or it doesn't work. I think it merits more testing before/if we decide to turn it on by default. Since it uses a setting, we can enable it on a per-application basis too.

It is possible to have a command that does context sensitive dictation in command mode, `dictate <user.prose>$: user.dictation_insert(prose)`. Should we add a command for this? I've been using `say` for it, but that is a big change in behavior from current knausj `say` (eg. if context-sensitivity is off, it will often make wrong guesses about capitalization & spacing; if context-sensitivity is on it does the clipboard dance a lot).

### CODE REVIEW

**Comment by knausj85** [comment]
Yeah, I also hesitate to globally enable this. Are there any well-behaved applications that should opt-in by default?

I personally use say as a poor-mans dictation mode, so I'm not opposed to changing its behavior. Presumably we would respect the setting too.

**Review by knausj85 — APPROVED** [review]

**Comment by rntz** [comment]
I think there are really two issues. One is application support and the other is clipboard flakiness. 

For application support, most text editors and web browsers will work fine, but with exceptions. Some web pages override default text editing behavior in strange ways, e.g. notion.so makes ctrl-c copy the surrounding block if nothing is selected, which screws us over; some text editors do a similar thing. Emacs doesn't work out-of-the-box because cancelling selections is different in emacs. Etc.

For flakiness, I know the clipboard is a bit flaky on Linux right now but I don't know what the state is on Mac/Windows, hopefully solid?

What I'd like to do is merge it disabled-by-default (which is the current default in the PR) and ask folks to test enabling it. Then we can selectively enable it by default in apps/contexts it works well in based on feedback. Once the draft window/text area stuff becomes more solid we can enable it there by default as well, I think that will be really nice.

**Comment by knausj85** [comment]
Makes sense - sounds reasonable.

**Comment by knausj85** [comment]
Oops, I haven't really encountered any issues with the clipboard stuff on Windows/Mac.

**Comment by knausj85** [comment]
I'm thinking to merge this alongside the draft ui -- either as a submodule or directly -- once 0.1.5 lands. The combo works very well.

**Comment by rntz** [comment]
I just had a video chat with someone on Mac where I had them try to use this in Gmail in a browser (not sure which one) on Mac, and it looked like they experienced some clipboard issue. Didn't have time to look into it but more reason to be cautious.

File: code/dictation.py [hunk 1]
```diff
@@ -1,102 +1,229 @@
...
+            # this will go right over the newline. Argh.
+            actions.edit.right()
+        return text
 
-    def auto_format_resume():
-        """Resumes the autoformatter"""
-        return auto_formatter.pause(False)
+    def clobber_selection_if_exists():
+        """Deletes the currently selected text if it exists; otherwise does nothing."""
+        actions.key("space backspace")
```
Comment (knausj85) [inline code/dictation.py:189]: A lot of the flakiness I'm observing in various applications is not actually the clipboard, but the fact that talon presses keys so fast. 

peek left is basically equivalent to
`actions.key("space backspace shift-alt-left shift-alt-left")` on Mac. This reliably fails in applications like Messages.

Potential fixes:
1.  check whether there's a selection before clobber, which is slow and may fail under load
2. Add a delay here. Fragile.
3. Go nuclear: we could consider making this context sensitive dictation only available in the draft window.

Comment (knausj85) [reply code/dictation.py:189]: 4. Tinker with the talon's key_wait stuff per-app

File: code/dictation.py [hunk 2]
```diff
@@ -1,102 +1,229 @@
...
-            if self.caps:
-                word = word.capitalize()
-
-            result += word
-            self.space = "\n" not in word and word[-1] not in no_space_after_these
-            self.caps = is_sentence_end
 
-        return result
+    def reset(self, before=None):
+        self.before = ""
```
Comment (rntz) [inline code/dictation.py:99]: Why is there a `before` parameter if it's not being used? Looks like there are no call sites that pass it in, so might as well remove the parameter.

**Comment by knausj85** [comment]
@dwiel @fidgetingbits - last call for concerns before I merge this on Sunday evening

**Pull request merged**